### DISCUSSION TITLE
CheckRules and IsDirty meta property and #issue no 945 (old forum) "Data Binding" for Rules

### DISCUSSION THREAD

**Discussion by theafricaninternetgroup** [MarimerLLC/csla#2128] [Questions] [answered]
i)I have a Blazor App which has page with a button bound to the IsSavable Property.When the PersonEdit is fetched from the db is the object runs BusinessRules.CheckRules() for the FullNameRule it turns that object is marked IsDirty as result of this call which also makes IsSavable true.This enables the save button on the page.The behaviour we want is for the save button to be enabled only for user initiated changes in this case editing the FirstName or LastName Properties
```
[Serializable]
public class PersonEdit :BusinessBase<PersonEdit>
{

 public static readonly PropertyInfo<string> IdProperty = RegisterProperty<string>(nameof(Id));
        public string Id
        {
            get => GetProperty(IdProperty);
            set => SetProperty(IdProperty, value);
        }
public static readonly PropertyInfo<string> FirstNameProperty = RegisterProperty<string>(nameof(FirstName));
        public string FirstName
        {
            get => GetProperty(FirstNameProperty);
            set => SetProperty(FirstNameProperty, value);
        }

        public static readonly PropertyInfo<string> LastNameProperty = RegisterProperty<string>(nameof(LastName));
        public string LastName
        {
            get => GetProperty(LastNameProperty);
            set => SetProperty(LastNameProperty, value);
        }
        public static readonly PropertyInfo<string> FullNameProperty = RegisterProperty<string>(nameof(FullName));
        public string FullName
        {
            get => GetProperty(FullNameProperty);
            private set => LoadProperty(FullNameProperty, value);
        }
        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new Csla.Rules.CommonRules.Dependency(FirstNameProperty, FullNameProperty));
            BusinessRules.AddRule(new Csla.Rules.CommonRules.Dependency(LastNameProperty, FullNameProperty));
            BusinessRules.AddRule(
                new FullNameRule
                {
                    PrimaryProperty = FullNameProperty
                });
        }

        private class FullNameRule : Csla.Rules.BusinessRule
        {
            protected override void Execute(IRuleContext context)
            {
                var target = context.Target as PersonEdit;
                var fullName = target.FirstName + " " + target.LastName;
                context.AddOutValue(PrimaryProperty, fullName);
            }
        }

        [Create]
        private void Create() { }
        [Fetch]
        private void Fetch(string id,[Inject]DataAccess.IPersonDal dal)
        {
            using (BypassPropertyChecks)
            {
                var data = dal.Fetch(id);
                this.FirstName = data.FirstName;
                this.LastName = data.LastName;
            }
            BusinessRules.CheckRules();
        }
    }
}

```
And the test we run is
```using System;
using Xunit;
using Csla;
using BusinessLib;
using DataAccess;
using Microsoft.Extensions.DependencyInjection;

namespace Test
{
    public class PersonEditTests
    {
        public PersonEditTests()
        {
            var services = new ServiceCollection();
            services.AddTransient<IPersonDal, PersonDal>();
            Csla.ApplicationContext.DefaultServiceProvider = services.BuildServiceProvider();
        }
        [Fact]
        public void FetchPersonEditTest()
        {
            var person = DataPortal.Fetch<PersonEdit>("1");

            //Assert
            Assert.False(person.IsSavable);
            Assert.False(person.IsDirty);
           
        }
    }
}


```
and the test failed
ii) I would also like to find out where issue **discussion # 945 (from the old forum) "Data Binding " for Rules** is going to be included in CSLA.NET 6

**Comment by jonnybee** [comment, marked as answer]
The issue here is that AddOutValue will make the object dirty if the value is changed from OuputProperty and so could trigger further rules on the changed property. 

I would suggest these changes: 

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new FullNameRule(FullNameProperty));
        }

        private class FullNameRule : Csla.Rules.BusinessRule
        {
            public FullNameRule(IPropertyInfo primaryProperty) : base(primaryProperty)
            {
                  // inputproperties is an implicit dependency - no need for explicit dependency rules 
                  InputProperties.Add(FirstNameProperty);
                  InputProperties.Add(LastNameProperty);
            }

            protected override void Execute(IRuleContext context)
            {
               // use implicit dependencies and let rule engine fetch values 
                var fullName = context.InputPropertyValues(FirstNameProperty) + " " + context.InputPropertyValues(LastNameProperty);
               // use load property to avoid making object dirty
                LoadProperty(context.Target, PrimaryProperty, fullName);
            }
        }

**Reply by bashinamwila** [reply]
Thanks @jonnybee  it worked though I still have warning that I should use atleast one Add method,not so much of any issue,Thanks alot

**Reply by bashinamwila** [reply]
@rockfordlhotka is this discussion/issue/feature going be included in CSLA.NET 6 **# 945 (from the old forum) "Data Binding " for Rules**

**Reply by rockfordlhotka** [reply]
@bashinamwila it is not currently (to my knowledge) in the backlog. If you don't see it in the Issues tab in the GitHub repo then it isn't on the plan.

Can you provide a link to the post you are talking about?

If you would like it to be considered, please write up what you think needs to happen and how it should work in a new issue and we can discuss. Please keep in mind that "data binding" varies from Windows Forms to Web Forms to MVC to XAML to Blazor, so (a) the problem may or may not exist for each UI framework; (b) the solution may vary depending on UI framework.
